File: ts-sdk/src/stg-session.ts
/**
 * stg（ステージングDB）の排他ロックと sync 元 revision 記録（設計 §15-11・TASK-55）。
 *
 * - `acquireStgLock`: `<stg>.lock` の排他ロック（PID ベース・新規依存なし）。書込系
 *   （sync / stg 編集セッション）が取得し、複数セッション/LLM の stg 同時編集を防止。
 * - `StgMeta` / `ProdRevision`: sync 時に prod の指紋を `<stg>.meta.json` に原子書き込みし、
 *   observe が prod の drift（sync 後更新）を検出して stale な promote をブロック。
 */
import fs from 'node:fs';
import path from 'node:path';
import type Database from 'better-sqlite3';

/** prod の revision 指紋（schema_version + 各テーブル件数/max-id・設計 §15-11）。 */
export interface ProdRevision {
  schemaVersion: number;
  mediaCount: number;
  mediaMaxId: number;
  tagsCount: number;
  tagsMaxId: number;
  mediaTagsCount: number;
  mediaAttributesCount: number;
  mediaHashesCount: number;
}

/** sync 元 prod 情報（revision 記録・設計 §15-11）。 */
export interface SyncedFrom {
  prodPath: string;
  revision: ProdRevision;
  syncedAt: string;
}

/** stg メタデータ（`<stg>.meta.json`）。 */
export interface StgMeta {
  syncedFrom: SyncedFrom;
}

/** stg が別セッションで使用中（排他ロック取得失敗・設計 §15-11）。 */
export class StgBusyError extends Error {
  readonly stgPath: string;
  readonly holderPid?: number;
  constructor(stgPath: string, holderPid?: number) {
    super(`Stg is busy (locked by another session): ${stgPath}`);
    this.name = 'StgBusyError';
    this.stgPath = stgPath;
    this.holderPid = holderPid;
  }
}

/** prod が別セッション/promote で使用中（prod 排他ロック取得失敗・設計 §5.3・TASK-58）。
 *  Rust `KijukuError::ProdBusy` と parity（StgBusy と分離）。 */
export class ProdBusyError extends Error {
  readonly prodPath: string;
  readonly holderPid?: number;
  constructor(prodPath: string, holderPid?: number) {
    super(`Prod is busy (locked by another promote/admin session): ${prodPath}`);
    this.name = 'ProdBusyError';
    this.prodPath = prodPath;
    this.holderPid = holderPid;
  }
}

/** promote gate 不合格（prod に触れる前に拒否・設計 §4.5/§6.4・TASK-58）。
 *  Rust `KijukuError::PromoteGateFailed` と parity。`failedChecks` は `${name}: ${detail}` 形式。 */
export class PromoteGateFailedError extends Error {
  readonly failedChecks: string[];
  constructor(failedChecks: string[]) {
    super(`Promote gate failed (prod not touched): ${failedChecks.join('; ')}`);
    this.name = 'PromoteGateFailedError';
    this.failedChecks = failedChecks;
  }
}

/** `<stgDbPath>.meta.json` のパス。 */
export function metaPath(stgDbPath: string): string {
  return `${stgDbPath}.meta.json`;
}

/** 2つの revision 指紋が一致するか（全フィールド比較・Rust の PartialEq と同等）。 */
export function prodRevisionEqual(a: ProdRevision, b: ProdRevision): boolean {
  return (
    a.schemaVersion === b.schemaVersion &&
    a.mediaCount === b.mediaCount &&
    a.mediaMaxId === b.mediaMaxId &&
    a.tagsCount === b.tagsCount &&
    a.tagsMaxId === b.tagsMaxId &&
    a.mediaTagsCount === b.mediaTagsCount &&
    a.mediaAttributesCount === b.mediaAttributesCount &&
    a.mediaHashesCount === b.mediaHashesCount
  );
}

/** `<stgDbPath>.lock` のパス。 */
export function lockPath(stgDbPath: string): string {
  return `${stgDbPath}.lock`;
}

/** 接続から prod revision 指紋を算出（5テーブルの集計 + schema_version）。 */
export function computeProdRevision(db: Database.Database): ProdRevision {
  const countMax = (table: string, idCol: string): [number, number] => {
    const row = db
      .prepare(`SELECT COUNT(*) AS c, COALESCE(MAX(${idCol}), 0) AS m FROM ${table}`)
      .get() as { c: number; m: number };
    return [Number(row.c), Number(row.m)];
  };
  const countOnly = (table: string): number => {
    const row = db.prepare(`SELECT COUNT(*) AS c FROM ${table}`).get() as { c: number };
    return Number(row.c);
  };

  const [mediaCount, mediaMaxId] = countMax('media', 'id');
  const [tagsCount, tagsMaxId] = countMax('tags', 'id');
  const mediaTagsCount = countOnly('media_tags');
  const mediaAttributesCount = countOnly('media_attributes');
  const mediaHashesCount = countOnly('media_hashes');
  const svRow = db
    .prepare('SELECT version FROM schema_version ORDER BY version DESC LIMIT 1')
    .get() as { version: number } | undefined;
  const schemaVersion = svRow ? Number(svRow.version) : 0;

  return {
    schemaVersion,
    mediaCount,
    mediaMaxId,
    tagsCount,
    tagsMaxId,
    mediaTagsCount,
    mediaAttributesCount,
    mediaHashesCount,
  };
}

/** stg メタを読込（ファイル不在時は undefined・後方互換）。 */
export function readStgMeta(stgDbPath: string): StgMeta | undefined {
  try {
    const content = fs.readFileSync(metaPath(stgDbPath), 'utf8');
    return JSON.parse(content) as StgMeta;
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') return undefined;
    throw e;
  }
}

/** stg メタを原子書き込み（tmp → rename・backup-meta パターンと同一）。 */
export function writeStgMeta(stgDbPath: string, meta: StgMeta): void {
  const p = metaPath(stgDbPath);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  const tmp = `${stgDbPath}.meta.json.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(meta, null, 2));
  fs.renameSync(tmp, p);
}

/** ロックファイルの内容（holder PID・取得時刻）。 */
interface LockFileContent {
  pid: number;
  acquiredAt: string;
}

/** PID が生存しているか（process.kill(0) で存在確認）。 */
function isPidAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

/** 排他ロックのハンドル。`release()` で解放。 */
export interface StgLock {
  readonly lockPath: string;
  release(): void;
}

/**
 * stg の排他ロックを取得（既に取得済みなら `StgBusyError`）。
 *
 * `<stg>.lock` を `O_EXCL`(wx) で作成し holder PID を書込む。既存時は holder PID を読み、
 * そのプロセスが既に終了していれば stale として破棄・再取得する。
 */
export function acquireStgLock(stgDbPath: string): StgLock {
  const lp = lockPath(stgDbPath);
  fs.mkdirSync(path.dirname(lp), { recursive: true });

  const createOrSteal = (): { fd: number } | { holder: number | undefined } => {
    try {
      const fd = fs.openSync(lp, 'wx');
      return { fd };
    } catch (e) {
      const code = (e as NodeJS.ErrnoException).code;
      if (code !== 'EEXIST') throw e;
    }
    // 既存ロック -> holder PID を読込
    let holderPid: number | undefined;
    try {
      const content = fs.readFileSync(lp, 'utf8').trim();
      const parsed = JSON.parse(content) as Partial<LockFileContent>;
      holderPid = typeof parsed.pid === 'number' ? parsed.pid : undefined;
    } catch {
      holderPid = undefined;
    }
    // stale（holder プロセス不在）なら破棄して再取得。PID 不明・生存中は StgBusy。
    if (holderPid !== undefined && !isPidAlive(holderPid)) {
      try {
        fs.unlinkSync(lp);
      } catch {
        /* ignore */
      }
      return createOrSteal();
    }
    return { holder: holderPid };
  };

  const result = createOrSteal();
  if (!('fd' in result)) {
    throw new StgBusyError(stgDbPath, result.holder);
  }
  const fd = result.fd;
  const content: LockFileContent = { pid: process.pid, acquiredAt: new Date().toISOString() };
  fs.writeFileSync(fd, JSON.stringify(content));

  let released = false;
  return {
    lockPath: lp,
    release(): void {
      if (released) return;
      released = true;
      try {
        fs.closeSync(fd);
      } catch {
        /* ignore */
      }
      try {
        fs.unlinkSync(lp);
      } catch {
        /* ignore */
      }
    },
  };
}

/**
 * prod の排他ロックを取得（promote/(b) 管理操作用・設計 §5.3・TASK-58）。
 *
 * 機構は `acquireStgLock` と同一（`${prod}.lock` の `O_EXCL` + holder PID stale 検出）だが、
 * 別セッション保持中は `StgBusyError` でなく `ProdBusyError` を throw する（Rust `ProdRwScope` と parity）。
 */
export function acquireProdLock(prodDbPath: string): StgLock {
  try {
    return acquireStgLock(prodDbPath);
  } catch (e) {
    if (e instanceof StgBusyError) {
      throw new ProdBusyError(prodDbPath, e.holderPid);
    }
    throw e;
  }
}

File: rust-sdk/src/stg_session.rs
//! stg（ステージングDB）の排他ロックと sync 元 revision 記録（設計 §15-11・TASK-55）。
//!
//! - `StgLock`: `<stg>.lock` の排他 advisory ロック（RAII・fs2）。書込系（sync / stg 編集
//!   セッション）が取得し、複数セッション/LLM の stg 同時編集を防止する。
//! - `StgMeta` / `ProdRevision`: sync 時に prod の指紋を `<stg>.meta.json` に原子書き込みし、
//!   observe が prod の drift（sync 後更新）を検出して stale な promote をブロックする。

use crate::error::{KijukuError, Result};
use crate::migration;
use chrono::{DateTime, Utc};
use fs2::FileExt;
use rusqlite::Connection;
use serde::{Deserialize, Serialize};
use std::fs::{self, File, OpenOptions};
use std::path::{Path, PathBuf};

// ========== revision 指紋 ==========

/// prod の revision 指紋（schema_version + 各テーブル件数/max-id・設計 §15-11）。
///
/// sync 元 prod のスナップショットを安価に表現する。完全な内容ハッシュではなく、
/// 行レベルの drift 検出用の「警告信号」であり、権威ある整合性判定は observe の
/// 実 diff + gate が担う。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProdRevision {
    pub schema_version: i64,
    pub media_count: i64,
    pub media_max_id: i64,
    pub tags_count: i64,
    pub tags_max_id: i64,
    pub media_tags_count: i64,
    pub media_attributes_count: i64,
    pub media_hashes_count: i64,
}

/// 接続から prod revision 指紋を算出（5テーブルの集計 + schema_version）。
pub fn compute_prod_revision(conn: &Connection) -> Result<ProdRevision> {
    let count_max = |table: &str, id_col: &str| -> Result<(i64, i64)> {
        let sql = format!(
            "SELECT COUNT(*), COALESCE(MAX({id}), 0) FROM {t}",
            id = id_col,
            t = table
        );
        let row: (i64, i64) = conn.query_row(&sql, [], |r| Ok((r.get(0)?, r.get(1)?)))?;
        Ok(row)
    };
    let count_only = |table: &str| -> Result<i64> {
        Ok(conn.query_row(&format!("SELECT COUNT(*) FROM {t}", t = table), [], |r| {
            r.get(0)
        })?)
    };

    let (media_count, media_max_id) = count_max("media", "id")?;
    let (tags_count, tags_max_id) = count_max("tags", "id")?;
    let media_tags_count = count_only("media_tags")?;
    let media_attributes_count = count_only("media_attributes")?;
    let media_hashes_count = count_only("media_hashes")?;
    let schema_version = migration::get_schema_version(conn)?;

    Ok(ProdRevision {
        schema_version,
        media_count,
        media_max_id,
        tags_count,
        tags_max_id,
        media_tags_count,
        media_attributes_count,
        media_hashes_count,
    })
}

// ========== メタデータサイドカー ==========

/// sync 元 prod 情報（revision 記録・設計 §15-11）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncedFrom {
    pub prod_path: String,
    pub revision: ProdRevision,
    pub synced_at: DateTime<Utc>,
}

/// stg メタデータ（`<stg>.meta.json`）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StgMeta {
    pub synced_from: SyncedFrom,
}

/// `<stg_db_path>.meta.json` のパス。
pub fn meta_path(stg_db_path: &Path) -> PathBuf {
    with_suffix(stg_db_path, ".meta.json")
}

/// `<stg_db_path>.lock` のパス。
pub fn lock_path(stg_db_path: &Path) -> PathBuf {
    with_suffix(stg_db_path, ".lock")
}

fn with_suffix(path: &Path, suffix: &str) -> PathBuf {
    let mut s = path.as_os_str().to_owned();
    s.push(suffix);
    PathBuf::from(s)
}

/// stg メタを読込（ファイル不在時は None・後方互換）。
pub fn read_stg_meta(stg_db_path: &Path) -> Result<Option<StgMeta>> {
    let path = meta_path(stg_db_path);
    match fs::read(&path) {
        Ok(bytes) => {
            let meta: StgMeta = serde_json::from_slice(&bytes)
                .map_err(|e| KijukuError::Parse(format!("stg meta parse error: {e}")))?;
            Ok(Some(meta))
        }
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(e.into()),
    }
}

/// stg メタを原子書き込み（tmp → rename・TASK-27 の backup-meta パターンと同一）。
pub fn write_stg_meta(stg_db_path: &Path, meta: &StgMeta) -> Result<()> {
    let path = meta_path(stg_db_path);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let json = serde_json::to_string_pretty(meta)
        .map_err(|e| KijukuError::Parse(format!("stg meta serialize error: {e}")))?;
    let tmp = with_suffix(stg_db_path, ".meta.json.tmp");
    fs::write(&tmp, &json)?;
    fs::rename(&tmp, &path)?;
    Ok(())
}

// ========== 排他ロック ==========

/// stg 排他ロック（advisory・設計 §15-11）。`<stg>.lock` の排他ロックを保持し、
/// `Drop` で解放（OS がプロセス終了/クラッシュ時にも自動解放・stale なし）。
///
/// ロックファイル自体は削除しない（unlink 競合で別プロセスのロックが無効化されるのを防ぐ）。
pub struct StgLock {
    file: File,
    path: PathBuf,
}

impl std::fmt::Debug for StgLock {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("StgLock").field("path", &self.path).finish()
    }
}

impl StgLock {
    /// stg の排他ロックを取得（既に取得済みなら `StgBusy`）。
    pub fn acquire(stg_db_path: &Path) -> Result<Self> {
        let path = lock_path(stg_db_path);
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }
        let file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .open(&path)?;
        file.try_lock_exclusive().map_err(|_| KijukuError::StgBusy {
            stg_path: stg_db_path.display().to_string(),
            holder_pid: None,
        })?;
        Ok(Self { file, path })
    }

    /// ロックファイルのパス。
    pub fn path(&self) -> &Path {
        &self.path
    }
}

impl Drop for StgLock {
    fn drop(&mut self) {
        // ベストエフォートで解放（失敗は無視・OS もプロセス終了時に解放する）。
        // ファイルは削除せず残す（unlink 競合回避）。
        let _ = self.file.unlock();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_stg_lock_acquire_then_reject_then_release() {
        let dir = TempDir::new().unwrap();
        let stg = dir.path().join("kijuku.stg.db");

        let lock1 = StgLock::acquire(&stg).expect("1回目の取得は成功");
        // 同一プロセスでも別 fd なので排他（flock は open file description 単位）
        let err = StgLock::acquire(&stg).unwrap_err();
        assert!(matches!(err, KijukuError::StgBusy { .. }), "2回目は StgBusy: {err:?}");

        drop(lock1);
        // 解放後は再取得可能
        let _lock2 = StgLock::acquire(&stg).expect("解放後の再取得は成功");
    }

    #[test]
    fn test_stg_meta_write_read_roundtrip() {
        let dir = TempDir::new().unwrap();
        let stg = dir.path().join("kijuku.stg.db");
        assert!(read_stg_meta(&stg).expect("read").is_none(), "不在時は None");

        let meta = StgMeta {
            synced_from: SyncedFrom {
                prod_path: "/tmp/prod.db".to_string(),
                revision: ProdRevision {
                    schema_version: 6,
                    media_count: 3,
                    media_max_id: 3,
                    tags_count: 0,
                    tags_max_id: 0,
                    media_tags_count: 0,
                    media_attributes_count: 0,
                    media_hashes_count: 0,
                },
                synced_at: Utc::now(),
            },
        };
        write_stg_meta(&stg, &meta).expect("write");
        let read = read_stg_meta(&stg).expect("read after write").expect("Some");
        assert_eq!(read.synced_from.revision.media_count, 3);
        assert_eq!(read.synced_from.prod_path, "/tmp/prod.db");
    }
}
